<?xml version="1.0"?>

<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="my_robot">

<xacro:macro name="r_link" params="prefix parent length radius rpy joint_xyz link_xyz">

    <!-- rpy (roll pitch yaw) describes rotation in radians -->
    <!-- xyz describes translation in meters -->
    <link name="${prefix}">
        <visual>
            <origin rpy="0 0 0" xyz="${link_xyz}"/> <!-- set the origin to 0 to recenter the link to the child_frame -->
            <geometry>
                <cylinder length="${length}" radius="${radius}"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="${link_xyz}"/> <!-- set the origin to 0 to recenter the link to the child_frame -->
            <geometry>
                <cylinder length="${length}" radius="${radius}"/>
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="${link_xyz}"/> <!-- set the origin to 0 to recenter the link to the child_frame -->
            <mass value="1"/>
            <inertia ixx="0.06" ixy="0.0" ixz="0.0" iyy="0.06" iyz="0.0" izz="0.25"/>
        </inertial>
    </link>


    <joint name="${parent}_to_${prefix}" type="continuous">
        <axis xyz="0 0 1"/>
        <parent link="${parent}"/>
        <child link="${prefix}"/>
        <origin rpy="${rpy}" xyz="${joint_xyz}"/> <!-- move the child_frame with respect to parent_frame -->
        <limit effort="30" velocity="1.0"/>
    </joint>


    <transmission name="${prefix}_transmission">
        <type>transmission_interface/SimpleTransmission</type>
        <actuator name="${prefix}_motor">
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
        <joint name="${parent}_to_${prefix}">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
    </transmission>


</xacro:macro>
</robot>

